base/reflection.jl

```diff
@@ -176,6 +176,14 @@ fieldnames(::Core.TypeofBottom) =
     throw(ArgumentError("The empty type does not have field names since it does not have instances."))
 fieldnames(t::Type{<:Tuple}) = ntuple(identity, fieldcount(t))
 
+"""
+    hasfield(T::DataType, name::Symbol)
+
+Returns a boolean indicating whether the DataType has the specified field as one of
+its own fields.
+"""
+hasfield(T::DataType, name::Symbol) = fieldindex(T, name, false) > 0
```

Comment:
Why does `hasproperty` take an object and `hasfield` takes a type?

Reply:
I supposed, [from this comment](https://github.com/JuliaLang/julia/issues/28850#issuecomment-415530388), that it needed to be that way. Should I make both refer to `DataType` instead @stevengj ?

Reply:
`hasproperty` *must* refer to an instance, *not* the type, similar to `propertynames`, because the properties may be determined at runtime from the instance.

Reply:
I was wondering if both should refer to instances. This would match `hasproperty` and `haskey`. Also, a `DataType` has fields of its own.

Reply:
But then hasfield would be equivalent to isdefined.

Reply:
I am trying to define `hasfield` as a `@pure` function. But it says `LoadError(at "compiler/compiler.jl" line 3: LoadError(at "reflection.jl" line 179: LoadError(at "reflection.jl" line 185: UndefVarError(var=Symbol("@pure")))))` I was just to make a `pure` function similar to https://github.com/JuliaLang/julia/blob/master/base/irrationals.jl#L93-L97

Can someone please tell me what I am missing?

Reply:
I think `reflection.jl` comes before `@pure` is defined.   Instead, just put `@_pure_meta` as the first line in the function, similar to several other examples in that file.